cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3836 @@
+                *(out + num_padding_zeros + num_digits - 1) = ('0' + x % 10);
+                num_digits -= 1;
+                x /= 10;
+            }
+        }
+
+        Result<std::shared_ptr<Schema>> SetOutputColumns(
+            const std::vector<std::string> &columns,
+            const std::vector<std::shared_ptr<DataType>> &types,
+            const std::unordered_map<std::string, int> &name_map,
+            std::vector<int> &gen_list)
+        {
+            gen_list.clear();
+            std::vector<std::shared_ptr<Field>> fields;
+            if(columns.empty())
+            {
+                fields.resize(name_map.size());
+                gen_list.resize(name_map.size());
+                for(auto pair : name_map)
+                {
+                    int col_idx = pair.second;
+                    fields[col_idx] = field(pair.first, types[col_idx]);
+                    gen_list[col_idx] = col_idx;
+                }
+                return schema(std::move(fields));
+            }
+            else
+            {
+                for(const std::string &col : columns)
+                {
+                    auto entry = name_map.find(col);
+                    if(entry == name_map.end())
+                        return Status::Invalid("Not a valid column name");
+                    int col_idx = static_cast<int>(entry->second);
+                    fields.push_back(field(col, types[col_idx]));
+                    gen_list.push_back(col_idx);
+                }
+                return schema(std::move(fields));
+            }
+        }
+
+        static TpchText g_text;
+
+        Status TpchText::InitIfNeeded(random::pcg32_fast &rng)
+        {
+            if(done_.load())
+                return Status::OK();
+
+            {
+                std::lock_guard<std::mutex> lock(text_guard_);
+                if(!text_)
+                {
+                    ARROW_ASSIGN_OR_RAISE(text_, AllocateBuffer(kTextBytes));
+                }
+            }
+            char *out = reinterpret_cast<char *>(text_->mutable_data());
+            char temp_buff[kChunkSize];
+            while(done_.load() == false)
+            {
+                int64_t known_valid_offset = 0;
+                int64_t try_offset = 0;
+                while(GenerateSentence(try_offset, rng, temp_buff))
+                    known_valid_offset = try_offset;
+
+                {
+                    std::lock_guard<std::mutex> lock(text_guard_);
+                    if(done_.load())
+                        return Status::OK();
+                    int64_t bytes_remaining = kTextBytes - generated_offset_;
+                    int64_t memcpy_size = std::min(known_valid_offset, bytes_remaining);
+                    std::memcpy(out + generated_offset_, temp_buff, memcpy_size);
+                    generated_offset_ += memcpy_size;
+                    if(generated_offset_ == kTextBytes)
+                        done_.store(true);
+                }
```

Comment:
```suggestion
                {
                    std::lock_guard<std::mutex> lock(text_guard_);
                    if(done_.load())
                        return Status::OK();
                    int64_t bytes_remaining = kTextBytes - generated_offset_;
                    int64_t memcpy_size = std::min(known_valid_offset, bytes_remaining);
                    int64_t start = out + generated_offset_;
                    generated_offset_ += memcpy_size;
                    if(generated_offset_ == kTextBytes)
                        done_.store(true);
                }
                std::memcpy(start, temp_buff, memcpy_size);
```
I doubt this would have a significant improvement to performance but it might be slightly faster.

Reply:
I've actually switched it to a lock-free implementation (with an atomic cursor), it seems to have sped up generation by quite a bit.

Reply:
Nevermind, it's the same speed between lock-free and locks (as far as I could tell, averages were the same but the lock free version had some more variance), so I just switched back to this version because it seems easier to understand.